### ISSUE TITLE
Dependency on Cloud Events

### ISSUE THREAD

**Issue by egekorkan** [w3c/wot-profile#126]
@mlagally has mentioned the F2F before TPAC of using cloud events in the profile and there are also other mentions of it in various issues. Something that needs to be documented somewhere is that the CloudEvents specification is driven by the Cloud Native Foundation, which is not a standards organization, thus cannot be simply linked to. This implies that having a dependency requires this profile spec to mirror their specification, as TD does with JSON Schema. The group should decide on whether this work is feasible in the current charter before deciding on adopting this specification.

**Comment by mlagally** [comment]
@egekorkan 
Thanks for flagging the problem. 
I think we do not have to reference the entire cloud event specification (which is not large btw.) but just select the parts that are relevant for us. As explained in #100, this is only a very small section, which could be just included in the same way.

**Comment by benfrancis** [comment]
Apart from not being from a standards body (whether in part or in full), I don't think a dependency on CloudEvents is needed or desirable for the Core Profile, see #100.

**Comment by mlagally** [comment]
@benfrancis 
I agree, we don't need the full cloud events, however a structured JSON payload with clear properties to be useful. Tying these into specifics of the the protocol binding is not desirable.
I think the event object should be protocol agnostic. 

Wrt. reference problem: we can resolve by defining a compatible payload as part of the profile.

**Comment by benfrancis** [comment]
> Tying these into specifics of the the protocol binding is not desirable.
>  think the event object should be protocol agnostic.

Again, "tying into specifics of the protocol" is the whole point of a protocol binding. Protocol bindings should not be protocol agnostic because their purpose is to be protocol specific. Other event mechanisms in other protocols may also have their own payload formats and may use a non-JSON serialisation or binary data format, so there's not even any guarantee that the same JSON wrapper could be re-used.

The current event protocol binding provides a very simple and neat mapping from the WoT information model onto the server-sent events event stream format:

1. The event name maps onto the `event` field
2. The data schema maps onto the `data` field
3. A timestamp can optionally be provided in the `id` field

Example Thing Description:
```
  "events": {
    "overheated": {
      "data": {
        "type": "number"
      },
      "forms": [{
        "href": "./events/overheated",
        "subprotocol": "sse"
      }]
    }
  },
```

Example event payload:
```
event: overheated\n
data: 90\n
id: 2021-11-16T16:53:50.817Z\n\n
```

Why overcomplicate this by adding in an unnecessary data wrapper which must then be added to the data schema of every single event affordance? Other than the non-goal of protocol agnosticism, what value does it provide to justify the added complexity?

**Comment by egekorkan** [comment]
Summary of my points across multiple issues and PRs since this issue is going to be used for discussions:
1. The [PR for CloudEvent](https://github.com/w3c/wot-profile/pull/235) was created 21 days ago and merged 22 days ago, meaning only for 1 day it was online. It took over work from https://github.com/w3c/wot-profile/pull/198 which had comments that were not resolved. PR #235 also got no reviews on GitHub. The [minutes](https://www.w3.org/2022/06/29-wot-profile-minutes.html) show that only Michael McCool, Michael Lagally, Kaz, and Mizushima-san were in the meeting and not the people who actually expressed opinions on this feature (me, Ben and Sebastian).
2. I think it is fundamentally wrong to propose this as mentioned also in https://github.com/w3c/wot-profile/issues/134#issuecomment-981722218. I am not against CloudEvents spec in any way, it is very important work. However, it makes zero sense in a world where TDs exist. CloudEvents simply adds a metadata layer around the actual payload. Most of that metadata is already in the TD as pointed out by @sebastiankb as well. More specifically, the payload format defines the following points, I am putting my notes next to each:
- specversion : can be put in the TD
- id: can remain, it makes sense
- source: base is already in the TD
- subject: this can be in the `href` of the TD
- type: even the text says copy content from TD
- time: can remain, it makes sense
- datacontenttype: we already have `contentType` in the TD
- data: we have `data` in the TD. Also, the fact that this is optional, conflicts  with :
  - https://w3c.github.io/wot-profile/#example-43 which only has a boolean data schema
  - some discussion in the TD(?) spec saying that we should not have empty payloads. Making this field optional means that we encourage empty payloads here. I am aware that the HTTP message will have a body but the other fields are basically header information put into the body.

Adding this information in a TD makes for longer TDs (maybe not depending on the my comment below) and bigger payloads. 

3. It is not clear what the consumer does with this information, what is the motivation behind it? From my point of view, once this event payload comes to the Consumer, it does not need a TD anymore.
4. I think seeing some TDs would make my point above even more clear. For example, the TD from Oracle at https://github.com/w3c/wot-testing/blob/main/events/2022.06.Online/Profiles/TD/Oracle/TDs/WoTWebThing.td.jsonld#L214 does not even contain this metadata. This raises an important discussion about whether this payload structure should be represented in the TDs as well or are they just assumed.
5. As expressed in the meeting minutes and the first comment here, it is a dependency like JSON Schema, which is not properly standardized yet. It is also in the incubation phase at CNCF: https://www.cncf.io/projects/#incubating . What kind of guarantees do we have? Since their spec is copied over here, we will not be compatible with them (a major argument from @mmccool and @mlagally ?) if they do some changes, even without bumping the version.

**Comment by mlagally** [comment]
@egekorkan 
 As the description in #235 clearly states #235 is a rebased version of a previous MR that was available for received various review comments and questions: #198 
 #198 has been available since May 5th. and received review from you, Sebastian and you had raised some questions. There were no change requests.

 All questions have been answered in 
 https://github.com/w3c/wot-profile/pull/198.
 
The content of the MR has not changed, it has been just rebased to resolve merge conflicts,
so it is not unreviewed content, but just a rebased version of #198, which was available for almost two months.

**Comment by egekorkan** [comment]
I think my core comment is not addressed anywhere: https://github.com/w3c/wot-profile/pull/198#pullrequestreview-963542459

If we are not aiming for others to change/adapt to our views of WoT, the whole profile specification makes no sense. We want people to do a specific thing when doing a new WoT implementation. If not, we have the entire binding templates documents that are all about explaining how you can use existing technology or standard in a WoT way. Maybe this is just my mistinterpretation.

> We are talking about green field devices, but not about green field cloud systems.

We are talking about devices and sometimes about consumers. We do not have a clear statement about the cloud. Also, if we are talking about cloud like this, we are becoming a cloud-specific HTTP profile which means we need a renaming.

**Comment by mlagally** [comment]
@egekorkan :
> I find a fundamental problem with using cloud events in a profile that is meant for greenfield applications. It adds too much metadata to payload and does not make sense in a WoT specification.

I believe this is your main comment. "too much" and "does not make sense" are opinions, but what "too much"? What means "does not make sense"?

Do you have a counterproposal about a scalable message format that will be easily integrated into changing existing event mechanisms of cloud systems?

**Comment by mlagally** [comment]
Payload formats for WebHooks are deferred to Profile 1.1. / Profile 2.0.
